scrapy/loader/__init__.py

```diff
@@ -35,8 +35,10 @@ def __init__(self, item=None, selector=None, response=None, parent=None, **conte
         self._local_values = defaultdict(list)
         # Preprocess values if item built from dict
         # Values need to be added to item._values if added them from dict (not with add_values)
-        for field_name, value in item.items():
-            self._values[field_name] = self._process_input_value(field_name, value)
+        if isinstance(item, dict):
+            for field_name, value in item.items():
+                # Convert all single values to lists because of following output processors
+                self._add_value(field_name, value)
```

Comment:
I’m not familiar enough with the internals of `ItemLoader`, so this may be a stupid question: Why do we want `_process_input_value` applied to dictionaries? Shouldn’t we be assigning the values *as is* (but as lists), as it *seems* we do with item classes?

Reply:
Not sure I understand the question correctly. If the question is "why to preprocess dicts" - it was the solution to save missing items (https://github.com/scrapy/scrapy/issues/3804), because without preprocessing scrapy dropped values. If I didn't understand the question - please, add more details :slightly_smiling_face:

Reply:
I mean that, with the current implementation, input processors are applied to input dictionary values, but not to item objects. `_add_value` calls `_process_input_value`, which passes dictionary values through input processors.

I wonder if it wouldn’t be better, to get values from dictionaries into `_values`, to use something similar to `_add_values` that does _not_ apply input processing.

I may be getting things wrong, though, and even if this is an issue it would be an issue that already exists before this patch, so maybe it’s better to leave this topic for later, for a different issue or pull request.